### PR TITLE
Restructure end-to-end integration

### PR DESCRIPTION
**Pull request by KomalBhadani27**
- Moved end-to-end integration from **main.py** to **stampifier.py**
- Moved **classifier_and_summarizer_main.py** code to **stampifier.py** and refactored accordingly.
- Removed  **classifier_and_summarizer_main.py**

### CODE REVIEW

File: stampifier.py [hunk 1]
```diff
@@ -1,41 +1,65 @@
...
+
+        if not self._website.contents.content_list:
+            LOGGER.debug("No content extracted")
+            return ''
+
+        _classifier_and_summarizer_response = self.get_stampified_content()
+
+        if not _classifier_and_summarizer_response["is_stampifiable"]:
+            LOGGER.debug("Website is not stampifiable")
+            return ''
```
Comment (detel) [inline stampifier.py:53]:
```suggestion
            return None
```

File: stampifier.py [hunk 2]
```diff
@@ -1,41 +1,65 @@
...
+        """This method is the pipeline between all the modules"""
+
+        _extractor = extractor.Extractor(self._website.url)
+        self._website.set_contents(_extractor.extract_html())
+
+        LOGGER.debug(self._website.convert_to_dict())
+
+        if not self._website.contents.content_list:
+            LOGGER.debug("No content extracted")
+            return ''
```
Comment (detel) [inline stampifier.py:47]:
```suggestion
            return None
```

File: stampifier.py [hunk 3]
```diff
@@ -63,6 +87,23 @@ def get_stampified_content(self):
...
+            = ExtractorOutputPreprocessor(self._website.contents)
+        self.preprocessed_contents_dict \
+            = output_preprocessor.get_preprocessed_content_lists()
+        self.title_text_contents \
+            = self.preprocessed_contents_dict["titles"]
+        self.normal_text_contents \
+            = self.preprocessed_contents_dict["sentences"]
+        self.media_contents = self.preprocessed_contents_dict["media"]
+        self.embedded_contents \
+            = self.preprocessed_contents_dict["embedded_content"]
```
Comment (detel) [inline stampifier.py:105]: @telescopic Why did this dic and not class (as in `data_model/PreprocessedContents`)? Why not pass entire PreprocessedContents object to Classifier, and Summarizer?

File: stampifier.py [hunk 4]
```diff
@@ -1,41 +1,65 @@
...
-            = self.preprocessed_contents_dict["sentences"]
-        self.media_contents = self.preprocessed_contents_dict["media"]
-        self.embedded_contents \
-            = self.preprocessed_contents_dict["embedded_content"]
+    def stampify(self):
+        """Starts the stampification process"""
+
+        if not self._website.is_valid:
+            LOGGER.error("Invalid URL!")
+            return ''
```
Comment (detel) [inline stampifier.py:33]:
```suggestion
            return None
```

File: stampifier.py [hunk 5]
```diff
@@ -1,41 +1,65 @@
...
-        self.normal_text_contents \
-            = self.preprocessed_contents_dict["sentences"]
-        self.media_contents = self.preprocessed_contents_dict["media"]
-        self.embedded_contents \
-            = self.preprocessed_contents_dict["embedded_content"]
+    def stampify(self):
+        """Starts the stampification process"""
+
+        if not self._website.is_valid:
+            LOGGER.error("Invalid URL!")
```
Comment (detel) [inline stampifier.py:32]: Isn't it better to throw exception instead of returning?

File: stampifier.py [hunk 6]
```diff
@@ -1,41 +1,65 @@
-''' This module contains the main interface definition
-for the classifier and summarizer module
-'''
+"""This script creates Stampify"""
+
+import logging
+
 from classification.classifier_main import Classifier
+from data_models.website import Website
+from extraction import extractor
```
Comment (detel) [inline stampifier.py:7]:
```suggestion
from extraction.extractor import Extractor
```

File: stampifier.py [hunk 7]
```diff
@@ -1,41 +1,65 @@
...
+        if not self._website.is_valid:
+            LOGGER.error("Invalid URL!")
+            return ''
+
+        return self.__convert_website_to_stamp()
+
+    def __convert_website_to_stamp(self):
+        """This method is the pipeline between all the modules"""
+
+        _extractor = extractor.Extractor(self._website.url)
```
Comment (detel) [inline stampifier.py:40]:
```suggestion
        _extractor = Extractor(self._website.url)
```

File: main.py [hunk 8]
```diff
@@ -73,9 +35,5 @@ def convert_website_to_stamp(_website, maximum_pages):
...
     url, maximum_pages = get_user_input()
 
-    _website = Website(url)
-
-    if not _website.is_valid:
-        LOGGER.error("Invalid URL!")
-    else:
-        convert_website_to_stamp(_website, maximum_pages)
+    _stampifier = Stampifier(url, maximum_pages)
+    _stamp_file_path = _stampifier.stampify()
```
Comment (detel) [inline main.py:39]: If you throw errors in stampify(), you can catch it here and log it. I guess it is better to throw exception instead of returning empty string for invalid outputs.

File: stampifier.py [hunk 9]
```diff
@@ -84,6 +125,13 @@ def _summarize(self):
 
         self.stampified_pages = summarizer.get_summarized_content()
 
+    def _order_stamp_pages(self):
```
Comment (detel) [inline stampifier.py:128]: Is this change intentional? Revert if else.

File: stampifier.py (hunk 3 above)
Comment (telescopic) [reply stampifier.py:105]: Ah yeah we could actually. Raised an issue for this! Will raise a PR for this once this PR is merged.
Hope this is alright! Thanks for the suggestion!

File: stampifier.py (hunk 5 above)
Comment (KomalBhadani27) [reply stampifier.py]: Yes, It is. Thanks for suggestion! Updated accordingly. Let me know if any changes required in that!

File: stampifier.py (hunk 9 above)
Comment (KomalBhadani27) [reply stampifier.py]: Done! It was not intentional. It was because of some misplacement while moving functionalities from one file to another.

File: stampifier.py (hunk 7 above)
Comment (KomalBhadani27) [reply stampifier.py]: Updated!

File: main.py (hunk 8 above)
Comment (KomalBhadani27) [reply main.py]: Updated accordingly. Let me know if further changes are required!

File: stampifier.py (hunk 4 above)
Comment (KomalBhadani27) [reply stampifier.py]: Raised error instead of returning anything as per suggestion!

File: main.py [hunk 10]
```diff
@@ -73,9 +42,15 @@ def convert_website_to_stamp(_website, maximum_pages):
...
-    if not _website.is_valid:
-        LOGGER.error("Invalid URL!")
-    else:
-        convert_website_to_stamp(_website, maximum_pages)
+    _stampifier = Stampifier(url, maximum_pages)
+
+    try:
+        _stamp_file_path = _stampifier.stampify()
+    except InvalidUrlError as err:
+        LOGGER.debug(err.message, err.provided_url)
```
Comment (detel) [inline main.py:50]: include provided_url in message, so that only LOGGER.debug(err.message) would be sufficient.

File: main.py [hunk 11]
```diff
@@ -73,9 +42,15 @@ def convert_website_to_stamp(_website, maximum_pages):
...
+    except NoneTypeMarkupError as err:
+        LOGGER.debug(err.message)
+    except WebsiteNotStampifiableError as err:
+        LOGGER.debug(err.message, "Error received by:", err.failure_source)
+    except WebsiteConnectionError as err:
+        LOGGER.debug(err.message, err.provided_url, err.trace)
```
Comment (detel) [inline main.py:56]: If every error has message, then we should simply for message. Let trace, url, source be part of message.
If you think message is meant for something else, you can log_statement in base error.

Comment (detel) [reply main.py:56]: Something like this:
```suggestion
    try:
        _stamp_file_path = _stampifier.stampify()
    except Error as err:
        LOGGER.debug(err.message)
```

File: stampifier.py [hunk 12]
```diff
@@ -1,32 +1,76 @@
...
+        self._website = Website(self.url)
         self.is_stampifiable = False
         # object of type StampPages
         self.stampified_pages = None
 
+    def stampify(self):
+        """Starts the stampification process"""
+
+        if not self._website.is_valid:
+            raise InvalidUrlError("Provided Url is Invalid!",
```
Comment (detel) [inline stampifier.py:33]: "Provided Url is Invalid!" could be within InvalidUrlError
Only pass params which are required, url in this case.

Comment (detel) [reply stampifier.py:33]: Same for rest of the errors.

File: main.py (hunk 10 above)
Comment (KomalBhadani27) [reply main.py]: Updated. Thanks for the suggestion!

File: main.py (hunk 11 above)
Comment (KomalBhadani27) [reply main.py]: Updated! Let me know if the new changes resolve this.

File: stampifier.py (hunk 12 above)
Comment (KomalBhadani27) [reply stampifier.py]: Updated!

File: stampifier.py (hunk 3 above)
Comment (KomalBhadani27) [reply stampifier.py]: Resolving this conversation!

File: stampifier.py [hunk 13]
```diff
@@ -1,32 +1,75 @@
...
+            raise WebsiteNotStampifiableError(
+                message="No content extracted!",
+                failure_source="Extractor")
+
+        _classifier_and_summarizer_response = self.get_stampified_content()
+
+        if not _classifier_and_summarizer_response["is_stampifiable"]:
+            raise WebsiteNotStampifiableError(
+                message="Website cannot be stampified!",
+                failure_source="Summarizer")
```
Comment (detel) [inline stampifier.py:55]: We should throw it in Classifier class.
```
def classify(self):
        if not is_page_stampifiable(self):
            raise WebsiteNotStampifiableError(
                message="Website cannot be stampified!",
                failure_source="Classifier")

def is_page_stampifiable(self):
        return max(
                max(self.normal_text_count, self.title_text_count),
                self.media_count) \
            + self.embedded_content_count >= self.min_pages
```

File: stampifier.py [hunk 14]
```diff
@@ -1,32 +1,75 @@
...
+        self._website.set_contents(_extractor.extract_html())
+
+        LOGGER.debug(self._website.convert_to_dict())
+
+        if not self._website.contents.content_list:
+            raise WebsiteNotStampifiableError(
+                message="No content extracted!",
+                failure_source="Extractor")
+
+        _classifier_and_summarizer_response = self.get_stampified_content()
```
Comment (detel) [inline stampifier.py:50]: get_stampified_content should only return stamp_pages.
Since error will be raised in Classifier itself.

File: stampifier.py [hunk 15]
```diff
@@ -1,32 +1,75 @@
...
+
+        _extractor = Extractor(self._website.url)
+        self._website.set_contents(_extractor.extract_html())
+
+        LOGGER.debug(self._website.convert_to_dict())
+
+        if not self._website.contents.content_list:
+            raise WebsiteNotStampifiableError(
+                message="No content extracted!",
+                failure_source="Extractor")
```
Comment (detel) [inline stampifier.py:48]: This should be raised in Extractor. We can assume `self._website.contents.content_list` is correct from extraction/. If it is incorrect, raise error in Extractor itself.

File: extraction/extractor.py [hunk 16]
```diff
@@ -48,19 +49,19 @@ def extract_html(self):
...
+            raise WebsiteConnectionError(self.url)
 
         # Remove the comment to read local files for testing purpose
         # file = open('./test_html.html','r').read()
 
-        if file:
-            self.soup = BeautifulSoup(file, 'lxml')
-            self.clean_soup()
+        if not file:
+            raise NoneTypeMarkupError
```
Comment (detel) [inline extraction/extractor.py:59]: missing brackets maybe???
```suggestion
            raise NoneTypeMarkupError()
```

Comment (KomalBhadani27) [reply extraction/extractor.py:59]: My bad! I will update it.

File: stampifier.py (hunk 15 above)
Comment (KomalBhadani27) [reply stampifier.py:48]: Makes sense! Sure.

File: stampifier.py (hunk 14 above)
Comment (KomalBhadani27) [reply stampifier.py:50]: @telescopic

File: stampifier.py (hunk 13 above)
Comment (KomalBhadani27) [reply stampifier.py:55]: @telescopic

File: stampifier.py (hunk 14 above)
Comment (KomalBhadani27) [reply stampifier.py:45]: @telescopic Can you look into this once this PR is merged and place the exceptions accordingly?

Comment (telescopic) [reply stampifier.py:45]: Noted. I'll raise issues for the same.
